metric.go

```diff
@@ -51,15 +56,42 @@ func NewMetricFamily(logContext string, mc *config.MetricConfig, constLabels []*
 	}
 
 	return &MetricFamily{
-		config:      mc,
-		constLabels: constLabels,
-		labels:      labels,
-		logContext:  logContext,
+		config:       mc,
+		constLabels:  constLabels,
+		labels:       labels,
+		logContext:   logContext,
+		globalConfig: gc,
 	}, nil
 }
 
 // Collect is the equivalent of prometheus.Collector.Collect() but takes a Query output map to populate values from.
 func (mf MetricFamily) Collect(row map[string]interface{}, ch chan<- Metric) {
+	var userLabels []*dto.LabelPair
+
+	// TODO: move to func()
+	if mf.config.JsonLabels != "" && row[mf.config.JsonLabels].(string) != "" {
+		var jsonLabels map[string]string
+
+		err := json.Unmarshal([]byte(row[mf.config.JsonLabels].(string)), &jsonLabels)
+		// errors silently ignored for now
+		if err != nil {
+			log.Warningf("[%s] Failed to parse JSON labels returned by query - %s", mf.logContext, err)
+		} else {
+			userLabelsMax := int(math.Min(float64(len(jsonLabels)), float64(mf.globalConfig.MaxJsonLabels)))
+			userLabels = make([]*dto.LabelPair, userLabelsMax)
+
+			idx := 0
+			for name, value := range jsonLabels {
+				// limit labels
+				if idx >= userLabelsMax {
```

Comment:
You don't need `userLabelsMax`, you can compare directly with `mf.globalConfig.MaxJsonLabels`.
```suggestion
				if idx >= mf.globalConfig.MaxJsonLabels {
```

You'd need to initialize `userLabels` with:

```go
userLabels = make([]*dto.LabelPair, 0, mf.globalConfig.MaxJsonLabels)
```

and use `append` below:

```go
userLabels = append(userLabels, makeLabelPair(&mf, name, value))
```

but it's the canonical way to do it in Go.

Reply:
ok, accepted